examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+            cls_outputs.append(
+                tf.reshape(self.cls_head(feature), [N, -1, self.num_classes])
+            )
+        cls_outputs = tf.concat(cls_outputs, axis=1)
+        box_outputs = tf.concat(box_outputs, axis=1)
+        return tf.concat([box_outputs, cls_outputs], axis=-1)
+
+
+"""
+## Implementing a custom layer to decode predictions
+"""
+
+
+class DecodePredictions(tf.keras.layers.Layer):
+    """Decodes RetinaNet Predictions"""
+
+    def __init__(
+        self,
+        num_classes=80,
+        confidence_threshold=0.05,
+        nms_iou_threshold=0.5,
+        soft_nms_sigma=0.4,
+        max_detections_per_class=100,
+        max_detections=100,
+        box_variance=[0.1, 0.1, 0.2, 0.2],
+        **kwargs
+    ):
+        super(DecodePredictions, self).__init__(**kwargs)
+        self.num_classes = num_classes
+        self.confidence_threshold = confidence_threshold
+        self.nms_iou_threshold = nms_iou_threshold
+        self.max_detections_per_class = max_detections_per_class
+        self.max_detections = max_detections
+
+        self._anchor_box = AnchorBox()
+        self._box_variance = tf.convert_to_tensor(
+            [0.1, 0.1, 0.2, 0.2], dtype=tf.float32
+        )
+
+    def _decode_box_predictions(self, anchor_boxes, box_predictions):
+        boxes = box_predictions * self._box_variance
+        boxes = tf.concat(
+            [
+                boxes[:, :, :2] * anchor_boxes[:, :, 2:] + anchor_boxes[:, :, :2],
+                tf.math.exp(boxes[:, :, 2:]) * anchor_boxes[:, :, 2:],
+            ],
+            axis=-1,
+        )
+        boxes_transformed = convert_to_corners(boxes)
+        return boxes_transformed
+
+    def call(self, images, predictions):
+        image_shape = tf.cast(tf.shape(images), dtype=tf.float32)
+        anchor_boxes = self._anchor_box.get_anchors(image_shape[1], image_shape[2])
+        box_predictions = predictions[:, :, :4]
+        cls_predictions = tf.nn.sigmoid(predictions[:, :, 4:])
+        boxes = self._decode_box_predictions(anchor_boxes[None, ...], box_predictions)
+        boxes = tf.expand_dims(boxes, axis=2)
+        boxes = tf.tile(boxes, [1, 1, self.num_classes, 1])
+
+        return tf.image.combined_non_max_suppression(
+            boxes,
+            cls_predictions,
+            self.max_detections_per_class,
+            self.max_detections,
+            self.nms_iou_threshold,
+            self.confidence_threshold,
+            clip_boxes=False,
+        )
+
+
+"""
+## Implementing Smooth L1 loss and Focal Loss as keras custom losses
+"""
+
+
+class RetinaNetBoxLoss(tf.losses.Loss):
+    """Implements Smooth L1 loss"""
+
+    def __init__(self, delta):
+        super(RetinaNetBoxLoss, self).__init__(
+            reduction="none", name="RetinaNetBoxLoss"
+        )
+        self._delta = delta
+
+    def call(self, y_true, y_pred):
+        loss = tf.compat.v1.losses.huber_loss(
```

Comment:
Use `tf.keras.losses.Huber()` (instantiated in the constructor)

Reply:
`tf.keras.losses.Huber()` averages the loss across all the boxes, but here we need to sum it across all boxes.  `tf.compat.v1.losses.huber_loss` returns the raw loss values, which I can then sum up.

Reply:
@tanzhenyu does something similar in [keras-cv](https://github.com/tanzhenyu/keras-cv/blob/master/kerascv/layers/losses/ssd_loss_layer.py#L32-#L40) by reimplementing the huber loss function.

Reply:
Then do `tf.reduce_sum(tf.keras.losses.huber(...))`, or alternatively use `tf.keras.losses.Uber(reduction='sum')`

Reply:
@fchollet tf.keras.losses.Huber has a bug, i.e., if `y_pred` and `y_true` is shape [batch_size, n_boxes, 4], the return loss would be [batch_size] where it calculates the mean over `n_boxes` -- this is wrong, *even* if reduction is None, because reduction is for first dimension only IIUC.
I think that's why the author was trying to use compat.v1 loss